tools/src/bin/build-blame.rs

```diff
@@ -0,0 +1,357 @@
+                    blame_parents,
+                    path.clone(),
+                )?;
+                builder.insert(entry.name().unwrap(), entry_builder.write()?, entry.filemode())?;
+            }
+            _ => {
+                panic!("Unexpected entry kind {:?} found in tree for commit {:?} at path {:?}", entry.kind(), commit.id(), path);
+            }
+        };
+
+        path.pop();
+    }
+
+    Ok(())
+}
+
+fn main() {
+    env_logger::init();
+
+    let args: Vec<_> = env::args().collect();
+    let git_repo = Repository::open(&args[1]).unwrap();
+    let blame_repo = Repository::open(&args[2]).unwrap();
+    let use_cinnabar = env::var("CINNABAR").map(|v| v == "1").unwrap_or(false);
+    let blame_ref = env::var("BLAME_REF").ok().unwrap_or("HEAD".to_string());
+
+    info!("Reading existing blame map of ref {}...", blame_ref);
+    let mut blame_map = if let Ok(oid) = blame_repo.refname_to_id(&blame_ref) {
+        let (blame_map, _) = index_blame(&blame_repo, Some(oid));
+        blame_map
+    } else {
+        HashMap::new()
+    };
+
+    let mut refobj = blame_repo.find_reference(&blame_ref)
+        .ok()
+        .and_then(|r| r.resolve().ok());
+
+    let mut walk = git_repo.revwalk().unwrap();
+    walk.set_sorting(Sort::TOPOLOGICAL | Sort::REVERSE).unwrap();
+    walk.push(git_repo.refname_to_id(&blame_ref).unwrap()).unwrap();
+    let revs_to_process = walk
+        .map(|r| r.unwrap()) // walk produces Result<git2::Oid> so we unwrap to just the Oid
+        .filter(|git_oid| !blame_map.contains_key(git_oid))
+        .collect::<Vec<_>>();
+    let rev_count = revs_to_process.len();
+
+    info!("Transforming new commits...");
+
+    let mut rev_done = 0;
+    for git_oid in revs_to_process {
+        rev_done += 1;
+
+        let hg_rev = if use_cinnabar {
+            get_hg_rev(&git_repo, &git_oid)
+        } else {
+            // we don't support mapfiles any more.
+            None
+        };
+
+        info!("Transforming {} (hg {:?}) progress {}/{}", git_oid, hg_rev, rev_done, rev_count);
+        let commit = git_repo.find_commit(git_oid).unwrap();
+        let blame_parents = commit.parent_ids().map(|pid| blame_repo.find_commit(blame_map[&pid]).unwrap()).collect::<Vec<_>>();
+
+        let mut file_movement = HashMap::new();
+        if commit.parent_count() == 1 {
+            let mut movement = HashMap::new();
+            let mut diff = git_repo.diff_tree_to_tree(
+                Some(&commit.parent(0).unwrap().tree().unwrap()),
+                Some(&commit.tree().unwrap()),
+                None
+            ).unwrap();
+            diff.find_similar(Some(DiffFindOptions::new().copies(true).rename_limit(1000000))).unwrap();
+            for delta in diff.deltas() {
+                if !delta.old_file().id().is_zero() && !delta.new_file().id().is_zero() && delta.old_file().path() != delta.new_file().path() {
+                    movement.insert(delta.new_file().id(),
```

Comment:
nit: Indentation here looks a bit odd, maybe rustfmt?

Reply:
Ah yeah, ran rustfmt on the whole thing now.